File: DotNetGotchas/VBDN/ParamThreadSafety/Modified1/ThreadPassingParams/ALib/SomeClass.vb
'SomeClass.vb part of ALib.dll
Imports System.Threading

Public Class SomeClass

    Private Sub Method1(ByVal val As Integer)
        Console.WriteLine( _
            "Method1 runs on Thread {0} with {1}", _
            AppDomain.GetCurrentThreadId(), val)

        ' Some operation takes place here
    End Sub

    Private theValToUseByCallMethod1 As Integer

    Private Sub CallMethod1()
        Method1(theValToUseByCallMethod1)
    End Sub


    Public Sub DoSomething(ByVal val As Integer)
        ' Some operation...

        ' Want to call Method1 in different thread
        ' from here?
        theValToUseByCallMethod1 = val
        Dim aThread As New Thread(AddressOf CallMethod1)
        aThread.Start()

        ' Some operation...
    End Sub
End Class

File: DotNetGotchas/CSharp/ParamThreadSafety/Modified2/ThreadPassingParams/ALib/SomeClass.cs
//SomeClass.cs part of ALib.dll
using System;
using System.Threading;

namespace ALib
{
	public class SomeClass
	{
		private void Method1(int val)
		{
			// Some operation takes place here
			Console.WriteLine(
				"Method1 runs on Thread {0} with {1}",
				AppDomain.GetCurrentThreadId(), val);
		}

		private int theValToUseByCallMethod1;

		private void CallMethod1()
		{
			Method1(theValToUseByCallMethod1);
		}

		public void DoSomething(int val)
		{
			// Some operation...

			// Want to call Method1 in different thread
			// from here?
			theValToUseByCallMethod1 = val;
			new Thread(new ThreadStart(CallMethod1)).Start();

			// Some operation...
		}
	}
}
